enterprise/cmd/cody-gateway/internal/actor/dotcomuser/dotcomuser.go

```diff
@@ -0,0 +1,148 @@
+package dotcomuser
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+	"time"
+
+	"github.com/Khan/genqlient/graphql"
+	"github.com/gregjones/httpcache"
+	"github.com/sourcegraph/log"
+
+	"github.com/sourcegraph/sourcegraph/enterprise/cmd/cody-gateway/internal/actor"
+	"github.com/sourcegraph/sourcegraph/enterprise/cmd/cody-gateway/internal/dotcom"
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/codygateway"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// dotcom user gateway tokens are always a prefix of 4 characters (sgd_)
+// followed by a 64-character hex-encoded SHA256 hash
+const tokenLength = 4 + 64
+
+var (
+	defaultUpdateInterval = 24 * time.Hour
+)
+
+type Source struct {
+	log    log.Logger
+	cache  httpcache.Cache
+	dotcom graphql.Client
+}
+
+var _ actor.Source = &Source{}
+
+func NewSource(logger log.Logger, cache httpcache.Cache, dotComClient graphql.Client) *Source {
+	return &Source{
+		log:    logger.Scoped("dotcomuser", "dotcom user actor source"),
+		cache:  cache,
+		dotcom: dotComClient,
+	}
+}
+
+func (s *Source) Name() string { return string(codygateway.ActorSourceDotcomUser) }
+
+func (s *Source) Get(ctx context.Context, token string) (*actor.Actor, error) {
+	// "sgd_" is dotcomUserGatewayAccessTokenPrefix
+	if token == "" || !strings.HasPrefix(token, "sgd_") {
+		return nil, actor.ErrNotFromSource{}
+	}
+
+	if len(token) != tokenLength {
+		return nil, errors.New("invalid token format")
+	}
+
+	data, hit := s.cache.Get(token)
+	if !hit {
+		return s.fetchAndCache(ctx, token)
+	}
+
+	var act *actor.Actor
+	if err := json.Unmarshal(data, &act); err != nil || act == nil {
+		s.log.Error("failed to unmarshal actor", log.Error(err))
+
+		// Delete the corrupted record.
+		s.cache.Delete(token)
+
+		return s.fetchAndCache(ctx, token)
+	}
+
+	if act.LastUpdated != nil && time.Since(*act.LastUpdated) > defaultUpdateInterval {
+		return s.fetchAndCache(ctx, token)
```

Comment:
I know you copied this `if` from the other source, but now I look at it, doesn't this mean that somehow the `LastUpdated` field is lost ... we stuck on never updating the `actor` again?

I think the more robust self-healing way is:

```suggestion
	if act.LastUpdated == nil || time.Since(*act.LastUpdated) > defaultUpdateInterval {
		return s.fetchAndCache(ctx, token)
```

@eseliger @bobheadxi WDYT?

Reply:
in the other source we also have periodic full sync which makes this less important. Since the design changed to full sync as the primary mechanism for product subscriptions, I haven't revisited this mechanism to make sure it works perfectly

Reply:
Makes sense to me.